AV1 Codec

jmillan opened the pull request:
Implements Dependency Descriptor (DD) RTP header extension and AV1 codec handler.

* DD Header extension is set as `recvonly` as we do not override it.

- [x] blocker https://github.com/versatica/mediasoup-client/pull/320

ibc commented:
> ```diff
> -                       pc.addTransceiver('video');
> +                       // Create video transceiver with simulcast in order to retrieve
> +                       // dependency descriptor header extension too.
> +                       pc.addTransceiver('video', {
> +                               sendEncodings: [{ rid: '0' }, { rid: 1 }],
> +                       });
> ```

Better than this, in Chrome 111 we can add a single encoding with `scalabilityMode`.

ibc commented on worker/src/RTC/Codecs/BitStream.cpp:
```diff
@@ -0,0 +1,52 @@
+{
+	namespace Codecs
+	{
+		BitStream::BitStream(const uint8_t* data, size_t len) : data(data), len(len)
```
Does it make sense that this class is inside Codecs? We may need it to parse whatever.

jmillan replied:
Yep, I was definitely thinking of moving it out. Maybe under `RTC` directly?

ibc replied:
I clearly see it in Utils namespace which is outside RTC.

ibc commented on worker/src/RTC/RtpDictionaries/RtpHeaderExtensionUri.cpp:
Aren't you removing frame-marking stuff in this PR?

jmillan replied:
I will, yes

ibc commented on worker/src/RTC/RtpPacket.cpp:
```diff
@@ -395,16 +395,17 @@ namespace RTC
+		this->rridExtensionId                 = 0u;
+		this->absSendTimeExtensionId          = 0u;
+		this->transportWideCc01ExtensionId    = 0u;
+		this->frameMarking07ExtensionId       = 0u;
+		this->frameMarkingExtensionId         = 0u;
```
Don't we want to remove everything related to frame-marking?

jmillan replied:
yes, yes

jmillan replied:
I'll remove frame marking in another PR as it's an orthogonal change that can be made even if this PR is not done.

ibc commented on worker/src/RTC/RtpDictionaries/RtpCodecMimeType.cpp:
```diff
@@ -35,6 +35,7 @@ namespace RTC
 		// Video codecs:
 		{ "vp8",             RtpCodecMimeType::Subtype::VP8             },
 		{ "vp9",             RtpCodecMimeType::Subtype::VP9             },
+		{ "av1",             RtpCodecMimeType::Subtype::AV1             },
```
Can we add it after h265 for consistency with `supportedRtpCapabilities.ts` and because this is the last added codec?

ibc commented on worker/src/RTC/RtpDictionaries/RtpCodecMimeType.cpp:
```diff
@@ -63,6 +64,7 @@ namespace RTC
 		// Video codecs:
 		{ RtpCodecMimeType::Subtype::VP8,             "VP8"             },
 		{ RtpCodecMimeType::Subtype::VP9,             "VP9"             },
+		{ RtpCodecMimeType::Subtype::AV1,             "AV1"             },
```
Same.

ibc reviewed (commented):
Partial super cosmetic review done for now.

jmillan committed:
worker: DD, rename Template by FameDependencyTemplate

jmillan committed:
cosmetic

jmillan committed:
AV1

jmillan committed:
worker: DD move BitStream to Utils

jmillan committed:
rust: DD RTP header extension

jmillan committed:
test: fix leak

jmillan committed:
remove Dump call for testing purposes

jmillan committed:
Move AV1 after H264_SVC

jmillan commented:
This PR is ready for review. Provides DD support and AV1 codec.

ibc commented on worker/src/RTC/Codecs/AV1.cpp:
```diff
@@ -0,0 +1,223 @@
+			payloadDescriptor->endOfFrame    = dependencyDescriptor->endOfFrame;
+			payloadDescriptor->spatialLayer  = dependencyDescriptor->spatialLayer;
+			payloadDescriptor->temporalLayer = dependencyDescriptor->temporalLayer;
+
```
Remove blank line?

ibc commented on worker/test/src/RTC/Codecs/TestDependencyDescriptor.cpp:
```diff
@@ -0,0 +1,100 @@
+	SECTION("parse Dependency Descriptor")
+	{
+		/**
+		 * Taken from https://issues.webrtc.org/issues/42225660.
```
Can we please use /** */ comment syntax with all the lines in the comment prefixed with *?

jmillan committed:
fuzzer

jmillan committed:
feedback: cosmetic

jmillan committed:
fuzzer: add missing files

ibc commented:
In fuzzer we must also check the AV1 codec, not just the dependency descriptor.

jmillan commented:
> In fuzzer we must also check the AV1 codec, not just the dependency descriptor.

I'll add it, but it does literally just assign the values from DD. There is no actual parsing.

jmillan committed:
worker: H264, enable Dependency Descriptor (#1517)

worker: H264, enable Dependency Descriptor

jmillan committed:
Fuzzer: AV1

jmillan committed:
format

jmillan commented:
> I'll add it, but it does literally just assign the values from DD. There is no actual parsing.

Done

jmillan committed:
fuzzer

Lynnworld commented:
note:  Dependency Descriptor  may use two byte extensions because it's too long. For example L3T3 mode

ibc commented:
> note:  Dependency Descriptor  may use two byte extensions because it's too long. For example L3T3 mode

This is already covered in mediasoup (2 byte extensions are supported since ever) and in mediasoup-client last version which negotiates it properly the way Chrome likes.

Lynnworld commented:
> > note:  Dependency Descriptor  may use two byte extensions because it's too long. For example L3T3 mode
> 
> This is already covered in mediasoup (2 byte extensions are supported since ever) and in mediasoup-client last version which negotiates it properly the way Chrome likes.

you missed something :( https://github.com/versatica/mediasoup/blob/09a23107e3e1020e4649ef95327902ab59c6d1fb/worker/src/RTC/Producer.cpp#L1423

ibc commented:
> you missed something :( https://github.com/versatica/mediasoup/blob/09a23107e3e1020e4649ef95327902ab59c6d1fb/worker/src/RTC/Producer.cpp#L1423

mediasoup doesn't add the Payload Descriptor header on packets it sends to clients. It only parses it from received packets.

Lynnworld commented:
For pipe consumer 😂

ibc commented:
>  For pipe consumer 😂

Excellent point

jmillan committed:
Worker: relay Dependency Descriptor RTP extension header

- Enable writting the frame number (thanks @alfredh for being an inspiration and example)

jmillan committed:
AV1: comment capability until ready

jmillan committed:
Merge branch 'v3' into dependency_descriptor

jmillan committed:
encoder

jmillan committed:
Revert "AV1: comment capability until ready"

This reverts commit 01ba26765eccca53c5817f04d9f31e228fc8247b.

jmillan committed:
create encoder

jmillan committed:
Merge remote-tracking branch 'origin/v3' into dependency_descriptor

jmillan committed:
lint

jmillan committed:
remove unneeded unique_ptr

jmillan committed:
PayloadDescriptorHandler: provide full RtpPacket

AV1 must modify the extension header rather than the payload.

jmillan committed:
fix test

jmillan committed:
fix node tests

jmillan committed:
Fuzzer: re-add DependencyDescriptor and AV1 files

jmillan committed:
format

jmillan committed:
format

jmillan committed:
AV1: add frameNumber to Dump()

jmillan committed:
add comment

jmillan committed:
Do not forward DD extension

When forwarding DD video freezes a lot with a 10% downlink loss in receiver.

https://issues.webrtc.org/issues/417857735

jmillan committed:
remove unused DD method

jmillan committed:
Merge remote-tracking branch 'origin/v3' into dependency_descriptor

jmillan committed:
CHANGELOG.md

jmillan committed:
fix node tests
